python/ray/serve/router.py

```diff
@@ -55,148 +57,158 @@ def __init__(
         """
-        for _ in range(len(self.in_flight_queries.keys())):
-            replica = next(self.replica_iterator)
-            if len(self.in_flight_queries[replica]
-                   ) >= self.max_concurrent_queries:
-                # This replica is overloaded, try next one
-                continue
-
-            logger.debug(f"Assigned query {query.metadata.request_id} "
-                         f"to replica {replica}.")
-            # Directly passing args because it might contain an ObjectRef.
-            tracker_ref, user_ref = replica.handle_request.remote(
-                query.metadata, *query.args, **query.kwargs)
-            self.in_flight_queries[replica].add(tracker_ref)
-            return user_ref
-        return None
-
-    @property
-    def _all_query_refs(self):
-        return list(
-            itertools.chain.from_iterable(self.in_flight_queries.values()))
-
-    def _drain_completed_object_refs(self) -> int:
-        refs = self._all_query_refs
-        done, _ = ray.wait(refs, num_returns=len(refs), timeout=0)
-        for replica_in_flight_queries in self.in_flight_queries.values():
-            replica_in_flight_queries.difference_update(done)
-        return len(done)
-
-    async def assign_replica(self, query: Query) -> ray.ObjectRef:
-        """Given a query, submit it to a replica and return the object ref.
-        This method will keep track of the in flight queries for each replicas
-        and only send a query to available replicas (determined by the backend
-        max_concurrent_quries value.)
+        with self.lock:
+            num_replicas = len(self.in_flight_queries.keys())
+            for _ in range(num_replicas):
+                replica = next(self.replica_iterator)
+                if len(self.in_flight_queries[replica]
+                       ) >= self.max_concurrent_queries:
+                    # This replica is overloaded, try next one
+                    continue
+
+                logger.debug(f"Assigned query {query.metadata.request_id} "
+                             f"to replica {replica}.")
+                # Directly passing args because it might contain an ObjectRef.
+                tracker_ref, user_ref = replica.handle_request.remote(
+                    query.metadata, *query.args, **query.kwargs)
+                tracker_ref._on_completed(
+                    functools.partial(
+                        self._on_query_tracker_completed,
+                        replica_handle=replica,
+                        tracker_ref=tracker_ref,
+                    ))
+                self.in_flight_queries[replica].add(tracker_ref)
+                future.set_result(user_ref)
+                return
+            raise RuntimeError(
+                "Postcondition failed. The replica set failed to assign "
+                f"query {query} to a replica.")
+
+    def assign_replica(self, query: Query) -> concurrent.futures.Future:
```

Comment:
This doesn't actually assign a replica anymore, should we call it something like `enqueue_query`?